### PR TITLE
Autoint++

### PR DESCRIPTION
**Pull request by D1MK4real**

### CODE REVIEW

File: lightautoml/text/embed.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+
+        Returns:
+            torch.Tensor
+
+        """
+        if not self._isfit:
+            raise RuntimeError("need to call `fit` or `from_summary` first")
+        X = X["cat"]
+        idxs: List[List[int]] = []
+        for row in X:
```
Comment (DESimakov) [inline lightautoml/text/embed.py:250]: Refactore code without rows loop. Use same logic as CatEmbedder.

File: lightautoml/text/embed.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+            X : Dict
+
+        Returns:
+            torch.Tensor
+
+        """
+        X = X["cont"]
+        if not self._isfit:
+            raise RuntimeError("need to call `fit` or `from_summary` first")
+        embedded = self.embedding_w.T.matmul(X.T.to(dtype=torch.float)).T + self.embedding_b
```
Comment (DESimakov) [inline lightautoml/text/embed.py:513]: It's OK, but use `.float()` instead of `.to(dtype=torch.float)`.

File: lightautoml/text/embed.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+
+
+class LinearEmbeddingFlat(LinearEmbedding):
+    """Flatten version of LinearEmbedding."""
+
+    def __init__(self, *args, **kwargs):
+        super(LinearEmbeddingFlat, self).__init__(*args, **{**kwargs, **{"flatten_output": True}})
+
+
+class DefaultEmbeddingFlat(DefaultEmbedding):
```
Comment (DESimakov) [inline lightautoml/text/embed.py:535]: Naming.

File: pyproject.toml [hunk 4]
```diff
@@ -48,7 +48,7 @@ catboost = ">=0.26.1"
 optuna = "*"
 torch = [
     {platform = "win32", python = "3.6.1", version = "1.7.0"},
-    {version = ">=1.0.0"}
```
Comment (DESimakov) [inline pyproject.toml:51]: Check min version with timm and transformers. (1.10?)

**Review by DESimakov — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 4 above)
Comment (D1MK4real) [reply pyproject.toml:51]: image

timm requirements

File: lightautoml/automl/presets/tabular_config.yml [hunk 5]
```diff
@@ -146,15 +152,15 @@ nn_params:
...
+  act_fun: LeakyReLU
   # add noise after dropout layer for more regularization
   use_noise: false
   # noise parameter
   noise_std: 0.05
   # use BatchNorm
   use_bn: true
   # define hidden layer dimensions for models in ['mlp', 'denselight', 'snn']
-  hidden_size: [512, 512, 512]
+  hidden_size: [512, 256, 128]
```
Comment (DESimakov) [inline lightautoml/automl/presets/tabular_config.yml:163]: Change to 2 hidden layers.

File: lightautoml/ml_algo/dl_model.py [hunk 6]
```diff
@@ -74,7 +87,58 @@
...
+    "cat": CatEmbedder,
+    "cat_no_dropout": BasicCatEmbeddingFlat,
+    "weighted": WeightedCatEmbeddingFlat,
+}
+cat_embedder_by_name = {"cat_no_dropout": BasicCatEmbedding, "weighted": WeightedCatEmbedding}
+cont_embedder_by_name_flat = {"cont": ContEmbedder, "linear": LinearEmbeddingFlat, "dense": DenseEmbeddingFlat}
+cont_embedder_by_name = {"linear": LinearEmbedding, "dense": DenseEmbedding}
+
+
+def _get_embedder_cat(params):
```
Comment (DESimakov) [inline lightautoml/ml_algo/dl_model.py:114]: Unnecessary functions. Use `.get `method in dicts.

File: lightautoml/ml_algo/torch_based/nn_models.py [hunk 7]
```diff
@@ -729,3 +790,189 @@ def __init__(self):
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward-pass."""
+        x = self.dense0(x)
+        x = self.features1(x)
+        x = self.features2(x)
+        return x.view(x.shape[0], -1)
+
+
+class AutoInt(nn.Module):
+    """The NODE model from https://github.com/Qwicen.
```
[the remark this answers is not in the document]
Comment (DESimakov) [reply lightautoml/ml_algo/torch_based/nn_models.py:860]: Still.

File: lightautoml/text/embed.py [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+            X : Dict
+
+        Returns:
+            torch.Tensor
+        """
+        X = X["cat"]
+        list_weights: List[List[List[float]]] = []
+        idxs_primary: List[List[int]] = []
+        idxs_default: List[List[int]] = []
+        for row in X:
```
Comment (DESimakov) [inline lightautoml/text/embed.py:334]: Rewrite two for loops.

File: lightautoml/text/nn_model.py [hunk 9]
```diff
@@ -270,9 +114,9 @@ def __init__(
         task: Task,
         torch_model: nn.Module,
         n_out: int = 1,
-        cont_embedder: Optional[Any] = None,
+        cont_embedder_: Optional[Any] = None,
```
Comment (DESimakov) [inline lightautoml/text/nn_model.py:117]: What is the purpose for additional post underscore?

File: pyproject.toml [hunk 10]
```diff
@@ -71,16 +71,16 @@ transformers = {version = ">=4", optional = true}
 
 # CV
 albumentations = {version = "<=1.0.3", optional = true}
-efficientnet-pytorch = {version = "*", optional = true}
+timm = {version = "*", optional = true}
```
[the remark this answers is not in the document]
Comment (DESimakov) [reply pyproject.toml:74]: Still.

File: pyproject.toml (hunk 4 above)
Comment (DESimakov) [reply pyproject.toml:51]: Still.

File: lightautoml/text/nn_model.py (hunk 9 above)
Comment (D1MK4real) [reply lightautoml/text/nn_model.py:117]: **params already has cont_embedder string field, so we need 1 more cont_embedder variable

**Review by DESimakov — APPROVED** [review]

**Review by DESimakov — APPROVED** [review]

**Pull request merged**